### PR TITLE
Add support for packages into quasiquotes and toolbox, improve handling of fresh names, unhardcode quasiquote expansion logic

### PR DESCRIPTION
**Pull request by densh**
Fixes:
- SI-6841 & SI-6657 by adding support for packages into quasiquotes and toolbox
- handling of fresh names that could have been generated by parser in quasiquotes
- unfortunate `isCaseDefStart` method name
- unhardcodes quasiquote expansion logic
- typos and various inconsistencies

Review by @retronym

### CODE REVIEW

**Comment by retronym** [comment]
This seems to need a rebase. Or you could just add a merge from master -> this branch, so you don't lose the test runs for the existing commits.

File: src/reflect/scala/reflect/internal/BuildUtils.scala [hunk 1]
```diff
@@ -119,6 +131,19 @@ trait BuildUtils { self: SymbolTable =>
 
     def RefTree(qual: Tree, sym: Symbol) = self.RefTree(qual, sym.name) setSymbol sym
 
+    def withFreshTermName[T](prefix: String)(f: TermName => T): T = f(TermName(freshName(prefix)))
+
+    def withFreshTypeName[T](prefix: String)(f: TypeName => T): T = f(TypeName(freshName(prefix)))
+
+    object freshName {
```
Comment (densh) [inline src/reflect/scala/reflect/internal/BuildUtils.scala]: This code duplicates subset of logic in `scala.tools.nsc.FreshNameCreator.Default`. AFAIK compilers' `FreshNameCreator` can't be used from reflection api not to depend on compiler.jar. Does anyone have any better ideas how to handle this?

Comment (retronym) [inline src/reflect/scala/reflect/internal/BuildUtils.scala]: You can refactor the compiler to delegate to this code in scala-reflect.jar.

**Comment by densh** [comment]
I've just rebased this pr on top of master. 

@retronym's feedback has been addressed in the following way:
1. isCaseDefEnd now accepts EOF in stock parser; surprisingly enough this improves error messages in some cases (a3c29c5)
2. FreshNameCreator now uses NameTransformer.encode for prefix sanitization (a42eef3)
3. FreshNameCreators are now created once per unit at compile-time and once per universe at run-time (c998145)
4. Internals of FreshNameCreator use ConcurrentHashMap and AtomicLong to prevent possible concurrency issues (c1fa3f1)

**Comment by densh** [comment]
@retronym is there anything else that needs to be fixed here?

File: src/compiler/scala/tools/reflect/quasiquotes/Parsers.scala [hunk 2]
```diff
@@ -54,7 +55,13 @@ trait Parsers { self: Quasiquotes =>
 
       def isHole(name: Name): Boolean = holeMap.contains(name)
 
+      override implicit def fresh = new FreshNameCreator {
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/quasiquotes/Parsers.scala]: annotate the return type here

**Comment by retronym** [comment]
Otherwise, LGTM. It's getting a bit late in the 2.11 cycle for too many more of these sort of changes though.

/cc @paulp who will probably have a branch to rebase on top of this and might therefore opine on these changes.

**Comment by gkossakowski** [comment]
This is insufficient commit message (f1b431d796dc873ae661d65bf9d45ae7459fb3cf):

> advanced fresh name reification

I don't understand what does this mean.

**Comment by gkossakowski** [comment]
> move fresh name creator into scala.reflect.internal.util

Why?

**Comment by densh** [comment]
**Q: What does advanced fresh name reification mean?**

A: Being macros, quasiquotes are on a different metalevel from the trees they are producing. So they have to transform trees into a trees that represents those trees (a.k.a. tree reification). E.g. during expansion of `q"foo"` code is first parsed into Ident(TermName("foo")) and then reified into:

```
Apply(Select(Select(Select(Select(Select(Ident(scala), scala.reflect), scala.reflect.runtime),
scala.reflect.runtime.package), TermName("universe")), TermName("Ident")), List(Apply(Select(
Select(Select(Select(Select(Ident(scala), scala.reflect), scala.reflect.runtime),
scala.reflect.runtime.package), TermName("universe")), TermName("TermName")),
List(Literal(Constant("foo"))))))
```

Which in the end is equivalent to following hand-written code:

```
scala.reflect.runtime.`package`.universe.Ident(scala.reflect.runtime.`package`.universe.TermName("foo"))
```

This includes reification of leaves of the trees like names (e.g. name is reified into a tree that represents call to TermName). During parsing some names are generated artificially using freshTermName & freshTypeName (e.g. `x$1`). Such names should be reified in a different way because they are assumed to be always fresh and non-overlapping with the environment. So `x$1` should reify down to equivalent of `freshTermName("x$")` rather than `TermName("x$1")`. 

But this is not enough. One name can be used more than once in a tree. E.g. `q"_ + 1"` desugars into `q"x$1 => x$1 + 1"`. So we need to ensure that every place where `x$1` is used gets the same fresh name. Hence the need for `withFreshTermName` that lets q"_ + 1" quasiquote desugare into equivalent of `withFreshTermName("x$") { freshx => q"$freshx => $freshx + 1" }`.

For pattern quasiquotes it's a bit different. Due to the fact that end-result must be a pattern we need to represent fresh names as patterns too. A natural way to express that something is fresh is to represent it as a free variable (e.g. any name will do in that place). But due to possible use of the same name in multiple places we need to make sure that all such places have the same values by adding a sequence of guards to the pattern.

Previously such names were reified naively and it could have caused name collision problems and inability to properly much on trees that contain such names. 

Implementation of this resides in `wrap`: https://github.com/scala/scala/commit/f1b431d796dc873ae661d65bf9d45ae7459fb3cf#diff-ffc63e6bc6012ef94c20797e29a4869dR30

I'll add this explanation to the commit message.

**Q: Why moving fresh name creator in internal util?**

A: Quasiquotes need to generate fresh names at runtime when withFreshTermName is called. Thus they need to use FreshNameCreator. Previously it was located in compiler jar and wasn't accessible from reflect jar (which led to code duplication) so it has to be moved to the reflect.

**Comment by gkossakowski** [comment]
@densh: thanks for excellent explanations. :+1: Please amend the second commit message as well. :)

**Comment by paulp** [comment]
Yes, this will savage my parser commit, but that's my problem. LGTM.

I saw all the changes to Token and it reminded me how nice it would be if this were a value class. It can't be until value classes don't inhibit switches from being generated.

**Comment by adriaanm** [comment]
>  It can't be until value classes don't inhibit switches from being generated.

that should be easy to fix

**Comment by paulp** [comment]
True that!

**Comment by adriaanm** [comment]
and it will, once I hit that rung of my TODO ladder

**Comment by som-snytt** [comment]
@paulp the opstack branch? It merges with barely a pin prick.
@adriaanm jacob's ladder?

**Comment by paulp** [comment]
@som-snytt no, the opstack code is a tiny carveoff of something approaching a parser rewrite. It's probably doomed anyway.

**Comment by densh** [comment]
Failure on 0fd119d looks spurious to me.

**Comment by densh** [comment]
PLS REBUILD/pr-scala@0fd119d

**Pull request merged**